hazelcast/src/test/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManagerTranslateTest.java

```diff
@@ -72,9 +71,124 @@ public void teardown() {
+        TcpClientConnectionManager clientConnectionManager =
+                new TcpClientConnectionManager(getHazelcastClientInstanceImpl(client));
+
+        clientConnectionManager.start();
+        clientConnectionManager.reset();
+
+        clientConnectionManager.getOrConnectToAddress(privateAddress);
+        provider.shouldTranslate = true;
+
+        // when
+        Connection connection = clientConnectionManager.getOrConnectToAddress(privateAddress);
+
+        // then
+        assertNotNull(connection);
+    }
+
+    @Test
+    public void testTranslateIsUsedWhenMemberHasPublicClientAddress() throws UnknownHostException {
+        // given
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED.getName(), "true");
+
+        HazelcastInstance client = HazelcastClientUtil.newHazelcastClient(null, clientConfig);
+        TcpClientConnectionManager clientConnectionManager =
+                new TcpClientConnectionManager(getHazelcastClientInstanceImpl(client));
+        clientConnectionManager.start();
+
+        // private member address is unreachable
+        Member member = new MemberImpl(new Address("192.168.0.1", 5701), VERSION, false, UUID.randomUUID());
+        // public member address is reachable
+        member.getAddressMap().put(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"),
+                new Address("127.0.0.1", 5701));
+
+        // when
+        Connection connection = clientConnectionManager.getOrConnectToMember(member);
+
+        // then
+        assertNotNull(connection);
+    }
+
+    @Test(expected = Exception.class)
+    public void testTranslateIsNotUsedWhenPublicIpDisabled() throws UnknownHostException {
+        // given
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.DISCOVERY_SPI_PUBLIC_IP_ENABLED.getName(), "false");
+
+        HazelcastInstance client = HazelcastClientUtil.newHazelcastClient(null, clientConfig);
+        TcpClientConnectionManager clientConnectionManager =
+                new TcpClientConnectionManager(getHazelcastClientInstanceImpl(client));
+        clientConnectionManager.start();
+
+        // private member address is incorrect
+        Member member = new MemberImpl(new Address("192.168.0.1", 5701), VERSION, false);
+        // public member address is correct
+        member.getAddressMap().put(EndpointQualifier.resolve(ProtocolType.CLIENT, "public"), new Address("127.0.0.1", 5701));
+
+        // when
+        clientConnectionManager.getOrConnectToMember(member);
+
+        // then
+        // throws exception because it can't connect to the incorrect member address
+    }
+
+    @Test(expected = Exception.class)
+    public void testTranslateFromMemberIsNotUsedWhenAlreadyTranslatedByAddressProvider() throws UnknownHostException {
```

Comment:
This test leads to a NPE (because `member.getUuid()` is `null`) which doesn't seems to be what it's supposed to be testing. Same story with `testTranslateIsNotUsedWhenPublicIpDisabled`.

Reply:
Fixed